src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlNormalizer.cs

```diff
@@ -433,7 +445,11 @@ internal override void Normalize(FieldInfo fi, object obj, Stream s)
         internal override void DeNormalize(FieldInfo fi, object recvr, Stream s)
         {
             byte[] b = new byte[4];
+#if NET8_0_OR_GREATER
+            s.ReadExactly(b, 0, b.Length);
```

Comment:
Am curious - ReadExactly and Read are very different things... If what you want here is ReadExactly (i.e. fill up b.Length bytes), doesn't that mean that pre-net8.0 has a bug? In other words, either in this point you really do want to loop and read repeatedly until b.Length bytes have been read (ReadExactly behavior), or you don't need that...

Reply:
If #2674 is any indication, this was added to fix the new warning CA2022. I think Read and ReadExactly will work identically in this context though: this is used to deserialize user-defined types from the database. The stream it's reading from will thus either be a MemoryStream (wrapped around the full byte value) or a custom stream which reads an entire field value at a time. In either event, the read should be able to completely populate the buffer.

It'd probably be a good idea to add a comment or some assertions on the return value of `s.Read` in netfx to be clearer here though...

Reply:
@edwardneal OK. I'm assuming that there are other places in SqlClient where  data is read from the network, and a loop is used to recreate ReadExactly behavior - you could also use that here, just in case (e.g. if somehow this code in the future runs against a stream that may not read fully).

In fact, I'd recommend simply having a ReadExactly extension method which does this ([see the Npgsql code](https://github.com/npgsql/npgsql/blob/main/src/Npgsql/Shims/StreamExtensions.cs)), for older TFMs. This way, you can simply use ReadExactly everywhere regardless of which TFM you're targeting, and don't need #if around your codebase.